provider

young1126 opened the pull request:
<img width="199" alt="image" src="https://user-images.githubusercontent.com/105976021/181744255-1ae8c108-2528-4acd-ab19-4ca754c82fcb.png">

direction page에 필요한 데이터를 가공하기 위해 loader 비스무리한거를 구현했습니다! 
데이터들을 다 읽어보니, http://jhshim.ddns.net/api/ulsanbus_arrivalinfos/?format=json 에서는
버스 노선 하나 당 하나의 버스에 대한 시간정보만 제공한다는 걸 깨달았습니다.
BusHeXA 웹에서 X역 전(YY정류장)과 같은 정보는 http://jhshim.ddns.net/api/posofbuses/?format=json 에서의 각 버스의 현재 node order에서 http://jhshim.ddns.net/api/nodeoflanes/?format=json 에서 각 버스의 울산과학기술원 정류장의 node order를 뺀 값을 이용하여 제공하는 것으로 추측했는데, 이를 어떤식으로 구현해야할지 모르겠습니당..
그리고 00:00 출발예정과 같은 정보는, 울산과학기술원이 기점인 버스들은 항상 이 정보만을 제공하고 있고,
울산과학기술원이 기점이 아닌 버스들의 경우에는 http://jhshim.ddns.net/api/posofbuses/?format=json 에서 학교로 오는 버스가 하나도 운행하지 않고 있을 시에 http://jhshim.ddns.net/api/ulsanbus_timetables/?format=json 에서 현재시간 과 가장 가까운 시간표를 제공해야한다는 것을 깨달았지만 어떻게 구현해야할지 모르겠습니다.

그리고 337버스가 특이한 케이스라는 걸 알았는데,
337(삼남신화 방면)과 337(태화강역 방면) 이 337(삼남 순환)의 도착 정보를 받아 쓰는 것 같아서 코드에서 이 버스들만 조금 다르게 처리하였습니다. 

그래서 프로바이더를 구현하기 위해
X분 후(YY정류장)만 제공하는 로더를 만들었습니다.
그래서 원래 형태에서 약간 바뀐
List<Map<String, dynamic>> = [{'bus_no' : String, 'info' : {'arrival_time' : int, 'current_node_name' : String, ~~~}}, {'bus_...}...]
로 하였습니다.

young1126 committed:
practiceprovider

directionpage를 위한 provider와 loader를 구현했습니다.

young1126 committed:
remove useless code

junwha commented on lib/direction_page.dart:
```diff
@@ -41,39 +42,42 @@ class _DirectionpageState extends State<Directionpage> {
   final _scrollController = ScrollController();
   @override
   Widget build(BuildContext context) {
+    var busInfo = Provider.of<Data>(context).updateData();
```
제 기억이 맞다면 여기서 필드를 가져와야 구독이 될거에요!
updateData 호출은 initState를 override한 함수에서 호출해주세요!

young1126 replied:
void initState() {    
    super.initState();
    Provider.of<Data>(context).updateData();
  }

이런식으로 하는건가요,,,? 계속 오류가 생겨요... ㅜㅜ

junwha commented on lib/direction_page.dart:
```diff
@@ -41,39 +42,42 @@ class _DirectionpageState extends State<Directionpage> {
+                child: const Text('정류소명 : 울산과학기술원', style: TextStyle(
+                  fontSize: 20, color: Colors.white)),
+              ),
+              ... Provider.of<Data>(context).busInfo.map((info) => buildBusinfo(info))
```
여기는 잘했어요!

junwha commented on lib/direction_page.dart:
```diff
@@ -124,4 +128,53 @@ Widget buildBusinfo(infoMap){
+              child: Column(
+                crossAxisAlignment: CrossAxisAlignment.start,
+                children: <Widget>[
+                  Container(
```
반복되는 위젯은 build 함수로 extract 해주세요!:)

young1126 replied:
buildbox 함수로 만들었습니다!

junwha commented on lib/loader/getAPI.dart:
```diff
@@ -0,0 +1,28 @@
+import 'dart:async';
```
이건 지호가 구현한 부분 가져온거 맞나요? 다음부터는 tracking이 쉽도록 merge 기능을 이용해주세요!

junwha commented on lib/main.dart:
```diff
@@ -10,8 +11,10 @@ class MyApp extends StatelessWidget {
-      home: LinePage(),
-      title: 'BusHeXA-line-page',    
+      title: 'BusHeXADirectionpage',
+      home:ChangeNotifierProvider<Data>(
```
잘 넣어줬는데 : 뒤에 공백 넣어주세요!

young1126 replied:
넵!

junwha commented on lib/provider/BusHeXAProvider.dart:
```diff
@@ -0,0 +1,21 @@
+
+
+class Data with ChangeNotifier{
+  List busInfo = [];
```
하나로 뭉뚱그리는 것보다는 changeNotifier에서 각 페이지에 각각 다른 필드를 뿌려주는게 좋을 것 같아요

young1126 replied:
혹시 최종적으로도 프로바이더 하나로 하는 건가용..? 전에 연습삼아 제 페이지만 해보라 하셔서 저렇게 만들었었어요 그럼 각페이지당 하나의 필드를 만들면 되나용..? (directionInfo, detailInfo, lineInfo, departureInfo)

junwha reviewed (commented):
전체적으로 많이 공부하고 잘 구현했네요!
리뷰 사항 반영하고 지호가 구현한 클래스로 교체하는 작업만 하면 어느 정도 마무리가 될 것 같아요:)

junwha commented on lib/loader/directionmodel.dart:
```diff
@@ -0,0 +1,38 @@
+  special['info'] = busInfomap[busNomap[1]];
+  dataList.add(special);
+
+  for (int i = 3; i < 10; i++) {
```
3부터 10까지 반복하는 이유가 뭔가요?

young1126 replied:
<img width="250" alt="image" src="https://user-images.githubusercontent.com/105976021/182079937-2cc57661-5723-4c51-bf25-c3bf4365274b.png">
웹버전에서보니 direction page에서 id 2(337삼남신화)부터 id9(233농소차고지) 순서대로 버스 정보가 나오길래 이렇게 처리를 했습니다. 2가 빠져있는 이유는, 2번의 정보가 전부 1번의 정보에서만 오고 실질적으로 2,3번은 정보가 안나오더라고요,,, 확인해보니 웹에서도 1번의 정보를 2번에 쓰고있는 것같길래 예외로 special 로 처리했습니다.

junwha commented:
PR에 나와있는 내용 중에 337에 해당하는 코드를 못 찾겠는데 혹시 지금 커밋된 코드가 latest 버전이 맞을까요?

junwha commented:
https://velog.io/@domini95/Flutter-Stateful-Widget-%EB%9D%BC%EC%9D%B4%ED%94%84-%EC%82%AC%EC%9D%B4%ED%81%B4-3.-initState

young1126 committed:
build box function

junwha committed:
Merge remote-tracking branch 'origin/dev' into impl-provider

young1126 committed:
08=15

junwha committed:
Merge remote-tracking branch 'origin/dev' into impl-provider

junwha committed:
Remove useless url dependency

junwha committed:
fix: change json decode part to dynamic

junwha committed:
Disable advanced DTO

junwha committed:
Reimplement direction model with OOP

- Define StopInfo class
- Change api calls to route key map
- Integrate all stop informations

junwha committed:
Enable test mode

junwha committed:
Add bus info to stop info

junwha committed:
Change key to info

junwha committed:
Change structure from map to list

junwha committed:
Remove unused classes

junwha committed:
fix: disable time info temporarily

junwha committed:
fix: get abs distance

junwha committed:
Modify build page

junwha committed:
Use only name for node

junwha committed:
Format codes

junwha committed:
Correct loader name

junwha committed:
Rename provider

junwha commented:
현재 정거장 정보가 불일치하는데 alias API를 활용해 별도 처리해줘야 할 것 같습니다.
- 337 예외처리
- alias 처리
- timetable 비교
순으로 작업이 필요합니다

junwha committed:
Filter negative stop left

- remove already passed bus

junwha committed:
Remove magic number

young1126 committed:
timeTable

junwha committed:
Extract base domain

junwha committed:
Implement timetable mechanism correctly

junwha committed:
Adjust threshold to filter non-department buses

junwha commented:
@young1126 아직 에러가 조금 있긴 한데 integration을 안 한지 너무 오래돼서 1차 검토 해주면 바로 머지 할게요!

junwha commented:
![image](https://user-images.githubusercontent.com/17183234/200452371-79a75389-6caf-48c7-b0c1-85b76d926ba0.png)
